### PR TITLE
add compression algorithms

### PR DESCRIPTION
**Pull request by tc80**
- [x] brotli11
- [x] gzip9

### CODE REVIEW

File: compress/algorithm.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		panic(fmt.Sprintf("%s failed: %s", alg, stdErr.String()))
+	}
+
+	return stdOut.Bytes()
+}
+
+// Brotli11 returns a brotli compressed file as bytes
+// at optimal compression (quality 11).
+func Brotli11(filePath string) []byte {
+	return runAlgorithm("brotli", "-c", "-q", "11", filePath)
```
Comment (klausenbusk) [inline compress/algorithm.go:33]: Yet another package which need to be installed. Should we document it somewhere, maybe in the README?

/ref #57

Comment (tc80) [reply compress/algorithm.go:33]: good point, will update README, thanks 😄

File: compress/algorithm.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// Gzip9 returns a gzip compressed file as bytes
+// at optimal compression (level 9).
+func Gzip9(filePath string) []byte {
+	return runAlgorithm("gzip", "-c", "-9", filePath)
```
Comment (klausenbusk) [inline compress/algorithm.go:39]: Wouldn't it be better to use the [gzip package](https://golang.org/pkg/compress/gzip/) instead of forking?

Comment (tc80) [reply compress/algorithm.go:39]: great point

File: README.md [hunk 3]
```diff
@@ -15,6 +15,12 @@ Repository that contains various tools for maintaining cdnjs.
 - `BOT_BASE_PATH`: cdnjs home
 - `SENTRY_DSN` sentry data source name (DSN)
 
+## Dependencies
+
+- [brotli](https://github.com/google/brotli)
+- [gzip](https://www.gzip.org/)
+- [zopflipng](https://github.com/google/zopfli)
+
```
Comment (klausenbusk) [inline README.md:23]:
```suggestion
- [jpegoptim](https://github.com/tjko/jpegoptim)
```
or
```suggestion
- [jpegoptim](https://www.kokkonen.net/tjko/projects.html)
```

I'm not sure if we should mention: `uglify-es`, `uglify-js` and `cleancss`. We should at least mention that you are expected to run `npm install` in the `tools` directory (I'm not sure why that approach was chosen, I would prefer just to run `npm install uglify-js -g` etc).

**Comment by klausenbusk** [comment]
Out of curiosity: What do we need this for? :)

**Comment by tc80** [comment]
> Out of curiosity: What do we need this for? :)

For [this](https://github.com/golang/go/issues/20075) reason. I was wondering why `gzip -d` and `brotli -d` were giving corruption error messages.

**Comment by klausenbusk** [comment]
> For [this](https://github.com/golang/go/issues/20075) reason

I meant, why do we need this PR. What do we need to compress? :)

**Comment by MattIPv4** [comment]
> I meant, why do we need this PR. What do we need to compress? :)

The old cdnjs auto update bot compressed image files — I assume this is going to be for implementing something similar to help reduce file sizes.

You may notice some packages have a dot file for not optimising images, this was from the old bot when it used compression, for cases when we didn’t want to compress images.

@tc80 correct me if I’m wrong on the intent behind this PR!

File: compress/algorithm.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package compress
+
+import (
+	"bytes"
+	"compress/gzip"
+
+	"github.com/andybalholm/brotli"
```
Comment (xtuc) [inline compress/algorithm.go:7]: I would prefer an official implementation, maybe call the brotli cli?

**Comment by xtuc** [comment]
cdnjs.cloudlare.com currently rely on Cloudflare on-the-fly compression which is good but want to increase the levels they are compressed at. Files will be smaller and that will reduce the bandwidth

**Pull request merged**